File: apps/kairos-web/src/app/services/market-data.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface MarketTick {
  time: string;
  symbol: string;
  exchange: string;
  price: number;
  volume: number;
}

export interface OhlcvCandle {
  bucket: string;
  symbol: string;
  exchange: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface PriceResponse {
  symbol: string;
  price: number | null;
  timestamp: string;
}

@Injectable({
  providedIn: 'root'
})
export class MarketDataService {
  private readonly apiUrl = `${environment.apiUrl}/api/market-data`;

  constructor(private http: HttpClient) {}

  /**
   * Get recent market ticks for a symbol
   */
  getRecentTicks(symbol: string, limit: number = 100): Observable<MarketTick[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<MarketTick[]>(`${this.apiUrl}/ticks/${symbol}`, { params });
  }

  /**
   * Get historical ticks within a time range
   */
  getHistoricalTicks(symbol: string, start: Date, end: Date): Observable<MarketTick[]> {
    const params = new HttpParams()
      .set('start', start.toISOString())
      .set('end', end.toISOString());
    return this.http.get<MarketTick[]>(`${this.apiUrl}/ticks/${symbol}/range`, { params });
  }

  /**
   * Get OHLCV candles
   */
  getOhlcvCandles(symbol: string, start?: Date, end?: Date, limit: number = 100): Observable<OhlcvCandle[]> {
    let params = new HttpParams();
    
    if (start && end) {
      params = params.set('start', start.toISOString()).set('end', end.toISOString());
    } else {
      params = params.set('limit', limit.toString());
    }
    
    return this.http.get<OhlcvCandle[]>(`${this.apiUrl}/ohlcv/${symbol}`, { params });
  }

  /**
   * Get latest price from DragonflyDB (real-time)
   */
  getLatestPrice(symbol: string): Observable<PriceResponse> {
    return this.http.get<PriceResponse>(`${this.apiUrl}/latest/${symbol}`);
  }
}

File: apps/kairos-api/src/main/java/com/kairos/service/MarketDataService.java
package com.kairos.service;

import com.kairos.model.MarketTick;
import com.kairos.model.OhlcvCandle;
import com.kairos.repository.MarketTickRepository;
import com.kairos.repository.OhlcvRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;

import java.time.Instant;

@Slf4j
@Service
@RequiredArgsConstructor
public class MarketDataService {
    
    private final MarketTickRepository marketTickRepository;
    private final OhlcvRepository ohlcvRepository;
    private final RealtimeDataService realtimeDataService;
    
    /**
     * Get recent market ticks from TimescaleDB
     */
    public Flux<MarketTick> getRecentTicks(String symbol, int limit) {
        log.debug("Fetching recent {} ticks for symbol: {}", limit, symbol);
        return marketTickRepository.findRecentBySymbol(symbol, limit)
            .doOnComplete(() -> log.debug("Fetched recent ticks for {}", symbol));
    }
    
    /**
     * Get historical market ticks within time range
     */
    public Flux<MarketTick> getHistoricalTicks(String symbol, Instant start, Instant end) {
        log.debug("Fetching historical ticks for {} from {} to {}", symbol, start, end);
        return marketTickRepository.findBySymbolAndTimeRange(symbol, start, end)
            .doOnComplete(() -> log.debug("Fetched historical ticks for {}", symbol));
    }
    
    /**
     * Get OHLCV candles from materialized view
     */
    public Flux<OhlcvCandle> getOhlcvCandles(String symbol, Instant start, Instant end) {
        log.debug("Fetching OHLCV candles for {} from {} to {}", symbol, start, end);
        return ohlcvRepository.findOhlcvBySymbol(symbol, start, end)
            .doOnComplete(() -> log.debug("Fetched OHLCV candles for {}", symbol));
    }
    
    /**
     * Get latest OHLCV candles
     */
    public Flux<OhlcvCandle> getLatestOhlcv(String symbol, int limit) {
        log.debug("Fetching latest {} OHLCV candles for {}", limit, symbol);
        return ohlcvRepository.findLatestOhlcv(symbol, limit)
            .doOnComplete(() -> log.debug("Fetched latest OHLCV for {}", symbol));
    }
    
    /**
     * Get latest price from DragonflyDB (real-time)
     */
    public reactor.core.publisher.Mono<Double> getLatestPrice(String symbol) {
        log.debug("Fetching latest price for {} from DragonflyDB", symbol);
        return realtimeDataService.getLatestPrice(symbol);
    }
}
